astroid/rebuilder.py

```diff
@@ -204,14 +214,73 @@ def _get_position_info(
             end_col_offset=t.end[1],
         )
 
+    def _fix_doc_node_position(self, node: NodesWithDocsType) -> None:
+        """Fix start and end position of doc nodes for Python < 3.8."""
+        if not self._data or not node.doc_node or node.lineno is None:
+            return
+        if PY38_PLUS:
+            return
+
+        lineno = node.lineno or 1  # lineno of modules is 0
+        end_range: Optional[int] = node.doc_node.lineno
+        if IMPLEMENTATION_PYPY:
+            end_range = None
+        data = "\n".join(self._data[lineno - 1 : end_range])
+
+        found_start, found_end = False, False
+        open_brackets = 0
+        skip_token: set[int] = {token.NEWLINE, token.INDENT}
+        if PY36:
+            skip_token.add(tokenize.NL)
+        else:
+            # token.NL was added in 3.7
+            skip_token.add(token.NL)
+
+        if isinstance(node, nodes.Module):
+            found_end = True
+            if PY36:
+                skip_token.add(tokenize.COMMENT)
+            else:
+                # token.COMMENT was added in 3.7
+                skip_token.add(token.COMMENT)
+
+        for t in generate_tokens(StringIO(data).readline):
+            if found_end is False:
+                if (
+                    found_start is False
+                    and t.type == token.NAME
+                    and t.string in {"def", "class"}
+                ):
+                    found_start = True
+                elif found_start is True and t.type == token.OP:
+                    if t.exact_type == token.COLON and open_brackets == 0:
+                        found_end = True
+                    elif t.exact_type == token.LPAR:
+                        open_brackets += 1
+                    elif t.exact_type == token.RPAR:
+                        open_brackets -= 1
+                continue
+            if t.type in skip_token:
+                continue
+            if t.type == token.STRING:
+                break
+            return
+        else:
+            return
+
+        node.doc_node.lineno = lineno - 1 + t.start[0]
```

Comment:
Why do we take `lineno - 1` here? Perhaps we should a comment explaining this? I can see us questioning this when we revisit this code at a later time.

Reply:
The token line number starts with 1 in the node line number. I.e. if we start tokenize in line 4, that's equivalent to `node.lineno`. However, for a match in line 6, the token position would be line 3, thus -1. `4 + 3 - 1 == 6`

I'll move the `-1` to the end, that should help.

Reply:
Ah understood! A comment might help still, but moving the `-1` will indeed help!

Reply:
> A comment might help still, but moving the `-1` will indeed help!

I'm a bit hesitant about the comment. It's a basic pattern that repeats itself whenever we do use tokenize in the `TreeRebuilder`. Should we add it to every return value? That feels unnecessary. Furthermore, there are multiple tests in place. So if you do end up removing it, you would know very quickly why it's necessary.

Reply:
Fine with me!